File: Cloude/Cloude/UI/YAMLParser.swift
import Foundation

enum YAMLParser {
    static func parse(_ text: String) -> Any? {
        let lines = text.components(separatedBy: "\n")
        let (result, _) = parseBlock(lines: lines, startIndex: 0, minIndent: 0)
        return result
    }

    private static func parseBlock(lines: [String], startIndex: Int, minIndent: Int) -> (Any?, Int) {
        var index = startIndex

        while index < lines.count {
            let line = lines[index]
            if line.trimmingCharacters(in: .whitespaces).isEmpty || line.trimmingCharacters(in: .whitespaces).hasPrefix("#") {
                index += 1
                continue
            }
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed.hasPrefix("- ") {
                return parseArray(lines: lines, startIndex: index, minIndent: indent(of: line))
            } else if trimmed.contains(":") {
                return parseDictionary(lines: lines, startIndex: index, minIndent: indent(of: line))
            }
            break
        }
        return (nil, index)
    }

    private static func parseDictionary(lines: [String], startIndex: Int, minIndent: Int) -> ([String: Any], Int) {
        var dict: [String: Any] = [:]
        var index = startIndex

        while index < lines.count {
            let line = lines[index]
            let lineIndent = indent(of: line)
            let trimmed = line.trimmingCharacters(in: .whitespaces)

            if trimmed.isEmpty || trimmed.hasPrefix("#") || trimmed == "---" {
                index += 1
                continue
            }

            if lineIndent < minIndent { break }
            if lineIndent > minIndent { break }

            guard let colonRange = trimmed.range(of: ":") else {
                index += 1
                continue
            }

            let key = String(trimmed[trimmed.startIndex..<colonRange.lowerBound]).trimmingCharacters(in: .whitespaces)
            let afterColon = String(trimmed[colonRange.upperBound...]).trimmingCharacters(in: .whitespaces)

            if afterColon.isEmpty {
                index += 1
                if index < lines.count {
                    let nextLine = lines[index]
                    let nextTrimmed = nextLine.trimmingCharacters(in: .whitespaces)
                    let nextIndent = indent(of: nextLine)
                    if !nextTrimmed.isEmpty && nextIndent > minIndent {
                        let (child, newIndex) = parseBlock(lines: lines, startIndex: index, minIndent: nextIndent)
                        dict[key] = child ?? ""
                        index = newIndex
                    } else {
                        dict[key] = ""
                    }
                }
            } else {
                dict[key] = parseScalar(afterColon)
                index += 1
            }
        }
        return (dict, index)
    }

    private static func parseArray(lines: [String], startIndex: Int, minIndent: Int) -> ([Any], Int) {
        var arr: [Any] = []
        var index = startIndex

        while index < lines.count {
            let line = lines[index]
            let lineIndent = indent(of: line)
            let trimmed = line.trimmingCharacters(in: .whitespaces)

            if trimmed.isEmpty || trimmed.hasPrefix("#") {
                index += 1
                continue
            }

            if lineIndent < minIndent { break }
            if lineIndent > minIndent { break }

            if trimmed.hasPrefix("- ") {
                let value = String(trimmed.dropFirst(2)).trimmingCharacters(in: .whitespaces)
                if value.contains(":") && !value.hasPrefix("\"") && !value.hasPrefix("'") {
                    let itemIndent = lineIndent + 2
                    let reconstructed = String(repeating: " ", count: itemIndent) + value
                    var subLines = [reconstructed]
                    var subIndex = index + 1
                    while subIndex < lines.count {
                        let subLine = lines[subIndex]
                        let subTrimmed = subLine.trimmingCharacters(in: .whitespaces)
                        if subTrimmed.isEmpty || subTrimmed.hasPrefix("#") {
                            subIndex += 1
                            continue
                        }
                        if indent(of: subLine) > lineIndent {
                            subLines.append(subLine)
                            subIndex += 1
                        } else {
                            break
                        }
                    }
                    let (child, _) = parseDictionary(lines: subLines, startIndex: 0, minIndent: itemIndent)
                    arr.append(child)
                    index = subIndex
                } else {
                    arr.append(parseScalar(value))
                    index += 1
                }
            } else {
                break
            }
        }
        return (arr, index)
    }

    private static func parseScalar(_ value: String) -> Any {
        if value == "true" || value == "True" || value == "yes" { return true }
        if value == "false" || value == "False" || value == "no" { return false }
        if value == "null" || value == "~" || value.isEmpty { return NSNull() }

        if (value.hasPrefix("\"") && value.hasSuffix("\"")) ||
           (value.hasPrefix("'") && value.hasSuffix("'")) {
            return String(value.dropFirst().dropLast())
        }

        if let intVal = Int(value) { return NSNumber(value: intVal) }
        if let doubleVal = Double(value) { return NSNumber(value: doubleVal) }

        if value.hasPrefix("[") && value.hasSuffix("]") {
            let inner = String(value.dropFirst().dropLast())
            return inner.components(separatedBy: ",").map { parseScalar($0.trimmingCharacters(in: .whitespaces)) }
        }

        return value
    }

    private static func indent(of line: String) -> Int {
        line.prefix(while: { $0 == " " }).count
    }
}

File: android/app/src/main/java/com/cloude/app/UI/files/YAMLParser.kt
package com.cloude.app.UI.files

object YAMLParser {
    fun parse(text: String): Any? {
        val lines = text.split("\n")
        val (result, _) = parseBlock(lines, 0, 0)
        return result
    }

    private fun parseBlock(lines: List<String>, startIndex: Int, minIndent: Int): Pair<Any?, Int> {
        var index = startIndex
        while (index < lines.size) {
            val line = lines[index]
            val trimmed = line.trim()
            if (trimmed.isEmpty() || trimmed.startsWith("#")) {
                index++
                continue
            }
            if (trimmed.startsWith("- ")) {
                return parseArray(lines, index, indentOf(line))
            } else if (trimmed.contains(":")) {
                return parseDictionary(lines, index, indentOf(line))
            }
            break
        }
        return null to index
    }

    private fun parseDictionary(lines: List<String>, startIndex: Int, minIndent: Int): Pair<Map<String, Any?>, Int> {
        val dict = mutableMapOf<String, Any?>()
        var index = startIndex

        while (index < lines.size) {
            val line = lines[index]
            val lineIndent = indentOf(line)
            val trimmed = line.trim()

            if (trimmed.isEmpty() || trimmed.startsWith("#") || trimmed == "---") {
                index++
                continue
            }

            if (lineIndent < minIndent || lineIndent > minIndent) break

            val colonIdx = trimmed.indexOf(':')
            if (colonIdx < 0) {
                index++
                continue
            }

            val key = trimmed.substring(0, colonIdx).trim()
            val afterColon = trimmed.substring(colonIdx + 1).trim()

            if (afterColon.isEmpty()) {
                index++
                if (index < lines.size) {
                    val nextLine = lines[index]
                    val nextTrimmed = nextLine.trim()
                    val nextIndent = indentOf(nextLine)
                    if (nextTrimmed.isNotEmpty() && nextIndent > minIndent) {
                        val (child, newIndex) = parseBlock(lines, index, nextIndent)
                        dict[key] = child ?: ""
                        index = newIndex
                    } else {
                        dict[key] = ""
                    }
                }
            } else {
                dict[key] = parseScalar(afterColon)
                index++
            }
        }
        return dict to index
    }

    private fun parseArray(lines: List<String>, startIndex: Int, minIndent: Int): Pair<List<Any?>, Int> {
        val arr = mutableListOf<Any?>()
        var index = startIndex

        while (index < lines.size) {
            val line = lines[index]
            val lineIndent = indentOf(line)
            val trimmed = line.trim()

            if (trimmed.isEmpty() || trimmed.startsWith("#")) {
                index++
                continue
            }

            if (lineIndent < minIndent || lineIndent > minIndent) break

            if (trimmed.startsWith("- ")) {
                val value = trimmed.removePrefix("- ").trim()
                if (value.contains(":") && !value.startsWith("\"") && !value.startsWith("'")) {
                    val itemIndent = lineIndent + 2
                    val reconstructed = " ".repeat(itemIndent) + value
                    val subLines = mutableListOf(reconstructed)
                    var subIndex = index + 1
                    while (subIndex < lines.size) {
                        val subLine = lines[subIndex]
                        val subTrimmed = subLine.trim()
                        if (subTrimmed.isEmpty() || subTrimmed.startsWith("#")) {
                            subIndex++
                            continue
                        }
                        if (indentOf(subLine) > lineIndent) {
                            subLines.add(subLine)
                            subIndex++
                        } else {
                            break
                        }
                    }
                    val (child, _) = parseDictionary(subLines, 0, itemIndent)
                    arr.add(child)
                    index = subIndex
                } else {
                    arr.add(parseScalar(value))
                    index++
                }
            } else {
                break
            }
        }
        return arr to index
    }

    private fun parseScalar(value: String): Any? {
        if (value == "true" || value == "True" || value == "yes") return true
        if (value == "false" || value == "False" || value == "no") return false
        if (value == "null" || value == "~" || value.isEmpty()) return null

        if ((value.startsWith("\"") && value.endsWith("\"")) ||
            (value.startsWith("'") && value.endsWith("'"))) {
            return value.drop(1).dropLast(1)
        }

        value.toLongOrNull()?.let { return it }
        value.toDoubleOrNull()?.let { return it }

        if (value.startsWith("[") && value.endsWith("]")) {
            val inner = value.drop(1).dropLast(1)
            return inner.split(",").map { parseScalar(it.trim()) }
        }

        return value
    }

    private fun indentOf(line: String): Int = line.takeWhile { it == ' ' }.length
}
